src/common/ceph_time.cc

```diff
@@ -328,3 +320,30 @@ std::chrono::seconds parse_timespan(const std::string& s)
 }
 
 }
+
+namespace std {
+template<typename Rep, typename Period>
+ostream& operator<<(ostream& m, const chrono::duration<Rep, Period>& t) {
+  if constexpr (chrono::treat_as_floating_point_v<Rep> ||
+                Period::den > 1) {
+    using seconds_t = chrono::duration<float>;
+    ::fmt::print(m, "{:.9}", chrono::duration_cast<seconds_t>(t));
+  } else {
+    ::fmt::print(m, "{}", t);
+  }
+  return m;
+}
+
+template ostream&
+operator<< <::ceph::timespan::rep,
+            ::ceph::timespan::period> (ostream&, const ::ceph::timespan&);
+
+template ostream&
+operator<< <::ceph::signedspan::rep,
+            ::ceph::signedspan::period> (ostream&, const ::ceph::signedspan&);
+
+template ostream&
+operator<< <chrono::seconds::rep,
+            chrono::seconds::period> (ostream&, const chrono::seconds&);
+
+} // namespace std
```

Comment:
This is Undefined Behavior, since we're adding our own overload to std.

Reply:
we were arguing on this before. see #23767 . for couple reasons, i think it's a compromise we need to live with:

- we rely on this undefined behavior already. i know, this is probably an excuse of a coward who does not dare to rework the whole tree. but i don't think it's worth doing so even we had enough bandwidth for this cleanup.
- it'd difficult to do this in a different way if we want to, for example, have something like https://github.com/ceph/ceph/blob/80cbe7ecaa23aefcd6432adf367312d9b5f0d7a3/src/test/common/test_time.cc#L148-L160

where gtest would like to print the timespan using `operator<<`.

Reply:
All right. Looking through the Google Test documentation it looks like their 'PrintTo' has to either be a member function (which we can't do) or in the same namespace (which is what I wanted to avoid) so there's no good workaround.

Reply:
@adamemerson thank you for looking into gtest for other alternatives!